core/src/main/scala/pureconfig/ConfigSource.scala

```diff
@@ -0,0 +1,284 @@
+
+/**
+ * A representation of a source from which `ConfigValue`s can be loaded, such as a file or a URL.
+ *
+ * A source allows users to load configs from this source as any type for which a `ConfigReader` is
+ * available. Raw configs can also be retrieved as a `ConfigValue`, a `ConfigCursor` or a
+ * `FluentConfigCursor`. Before using any of the loading methods described, Users can opt to focus
+ * on a specific part of a config by specifying a namespace.
+ *
+ * All config loading methods are lazy and defer resolution of references until needed.
+ */
+trait ConfigSource {
+
+  /**
+   * Retrieves a `ConfigValue` from this source.
+   *
+   * @return a `ConfigValue` retrieved from this source.
+   */
+  def value: Result[ConfigValue]
+
+  /**
+   * Returns a cursor for a `ConfigValue` retrieved from this source.
+   *
+   * @return a cursor for a `ConfigValue` retrieved from this source.
+   */
+  def cursor: Result[ConfigCursor] =
+    value.right.map(ConfigCursor(_, Nil))
+
+  /**
+   * Returns a fluent cursor for a `ConfigValue` retrieved from this source.
+   *
+   * @return a fluent cursor for a `ConfigValue` retrieved from this source.
+   */
+  def fluentCursor: FluentConfigCursor =
+    FluentConfigCursor(cursor)
+
+  /**
+   * Navigates through the config to focus on a namespace.
+   *
+   * @param namespace the namespace to focus on
+   * @return a new `ConfigSource` focused on the given namespace.
+   */
+  def at(namespace: String): ConfigSource =
+    ConfigSource.fromCursor(fluentCursor.at(PathUtil.splitPath(namespace).map(p => p: PathSegment): _*))
+
+  /**
+   * Loads a configuration of type `A` from this source.
+   *
+   * @tparam A the type of the config to be loaded
+   * @return A `Right` with the configuration if it is possible to create an instance of type
+   *         `A` from this source, a `Failure` with details on why it isn't possible otherwise
+   */
+  def load[A](implicit reader: Derivation[ConfigReader[A]]): Result[A] =
+    cursor.right.flatMap(reader.value.from)
+
+  /**
+   * Loads a configuration of type `A` from this source. If it is not possible to create an
+   * instance of `A`, this method throws a `ConfigReaderException`.
+   *
+   * @tparam A the type of the config to be loaded
+   * @return The configuration of type `A` loaded from this source.
+   */
+  @throws[ConfigReaderException[_]]
+  def loadOrThrow[A: ClassTag](implicit reader: Derivation[ConfigReader[A]]): A = {
+    load[A] match {
+      case Right(config) => config
+      case Left(failures) => throw new ConfigReaderException[A](failures)
+    }
+  }
+}
+
+/**
+ * A `ConfigSource` which is guaranteed to generate config objects (maps) as root values.
+ *
+ * @param config the thunk to generate a `Config` instance. This parameter won't be memoized so it
+ *               can be used with dynamic sources (e.g. URLs)
+ */
+class ConfigObjectSource(config: => Result[Config]) extends ConfigSource {
```

Comment:
We might want to make this `final` (and the constructor `private`) so we don't leak the implementation, and have the ability to add new methods later in a binary compatible way.